File: side_project_aoc/aoc_2017/aoc1701.py
path = '1701.0'
# path = '1701.1'
pfix = '_inputs_/'

def main():
    file = open(pfix + path).read().strip()
    a = [s for s in file.split('\n')]
    print(a)
    res, res2 = 0, 0
    for s in a:
        res, res2 = solver1(s), solver2(s)
    print('Star 1:', res)
    print('Star 2:', res2)

def solver2(s: str) -> int:
    res = 0
    S = len(s)
    mid = S // 2
    for i in range(S):
        if i + mid < S and s[i] == s[i + mid]: res += 2 * int(s[i])
    return res

def solver1(s: str) -> int:
    # Z = ord('0')
    S = len(s)
    res = 0
    for i in range(S):
        if (i != S-1 and s[i] == s[i + 1]) or (i == S-1 and s[i] == s[0]):
            res += int(s[i])
    return res

if __name__ == '__main__':
    main()

File: side_project_aoc/aoc_2017/aoc1701.cc
#include "iostream"
#include "vector"

#define end endl

using	namespace std;

int	solver1(string &);
int	solver2(string &);

int	main()
{
	vector<string>	a;
	string		s;
	int r1 = 0, r2 = 0;

	while (!cin.eof() && cin >> s)
		a.push_back( s );
	cout << "p1: ";
	for (string ss: a)
		r1 = solver1(ss);
	cout << "\np2: ";
	for (string ss: a)
		r2 = solver2(ss);
	cout << "\n\nStar 1: " << r1 << "\nStar 2: " << r2 << end;
}

int	solver2(string &s)
{
	int i = -1;
	int r = 0, len = s.length(), mid = len / 2;
	while (++i < len)
	{
		if (mid + i < len && s[i] == s[mid + i])
			r += (s[i] - '0') * 2;
	}
	cout << r << ' ';
	return r;
}

int	solver1(string& s)
{
	int i = -1;
	int r = 0;
	int len = s.length();
	while (++i < len)
	{
		if ((i ^ len - 1 && s[i] == s[i + 1])
		|| (i == len - 1 && s[i] == s[0]))
		{
			r += s[i] - '0';
		}
	}
	cout << r << ' ';
	return r;
}
